### PR TITLE
Add workflow to run detox iOS tests

### PR DESCRIPTION
**Pull request by yasserfaraazkhan**
#### Summary

- Added ios detox Workflow to run tests
- Running tests in parallel.
- Generates combined report and sends to channel.

#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: 

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: .github/workflows/e2e-detox.yml [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        id: generate-specs
+        uses: ./.github/actions/generate-specs
+        with:
+          parallelism: 10
+          directory: detox/e2e
+          search_path: test
+          device: ${{ env.DEVICE_NAME }}
+          os_version: ${{ env.DEVICE_OS_VERSION }}
+
+  # build-ios-simulator:
```
Comment (larkox) [inline .github/workflows/e2e-detox.yml:69]: I see several sections commented in this file. Are they something you will work on later, are they going to be addressed in this PR, or can they be removed?

File: detox/e2e/support/test_config.ts [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
+export const siteTwoUrl = process.env.SITE_2_URL || 'https://mobile-e2e-site-2.test.mattermost.cloud';
+export const serverThreeUrl = process.env.SITE_3_URL || 'https://mobile-e2e-site-3.test.mattermost.cloud';
+export const siteThreeUrl = process.env.SITE_3_URL || 'https://mobile-e2e-site-3.test.mattermost.cloud';
 export const smtpUrl = process.env.SMTP_URL || 'http://127.0.0.1:9001';
-export const adminEmail = process.env.ADMIN_EMAIL || 'sysadmin@sample.mattermost.com';
-export const adminUsername = process.env.ADMIN_USERNAME || 'sysadmin';
-export const adminPassword = process.env.ADMIN_PASSWORD || 'Sys@dmin-sample1';
+export const adminEmail = process.env.ADMIN_EMAIL || 'success+sysadmin@simulator.amazonses.com';
+export const adminUsername = process.env.ADMIN_USERNAME || 'admin';
+export const adminPassword = process.env.ADMIN_PASSWORD || 'admin@suz4ziqx9jbnjx4bwmfitfknpa';
```
Comment (larkox) [inline detox/e2e/support/test_config.ts:13]: Shouldn't these be passed by environment variables? Is any information here information should not be made public?

File: detox/e2e/support/ui/screen/create_or_edit_channel.ts [hunk 3]
```diff
@@ -59,6 +59,8 @@ class CreateOrEditChannelScreen {
 
     openEditChannel = async () => {
         // # Open edit channel screen
+        await ChannelInfoScreen.scrollView.tap({x: 1, y: 1});
+        await ChannelInfoScreen.scrollView.scroll(100, 'down');
```
Comment (larkox) [inline detox/e2e/support/ui/screen/create_or_edit_channel.ts:63]: Why is this needed? Shouldn't it open the edit screen just right without this?

File: detox/e2e/support/ui/screen/login.ts [hunk 4]
```diff
@@ -60,9 +60,19 @@ class LoginScreen {
 
     login = async (user: any = {}) => {
         await this.toBeVisible();
-        await this.usernameInput.replaceText(user.username);
-        await this.passwordInput.replaceText(user.password);
-        await this.signinButton.tap();
+        await this.usernameInput.typeText(`${user.newUser.email}\n`);
+        await this.passwordInput.typeText(`${user.newUser.password}\n`);
+
+        // await this.signinButton.tap();
```
Comment (larkox) [inline detox/e2e/support/ui/screen/login.ts:66]: If we don't tap the button, we shouldn't be able to log in, right? Or the newline in line 64 does the trick?

File: detox/e2e/test/account/about.e2e.ts [hunk 5]
```diff
@@ -69,7 +69,6 @@ describe('Account - Settings - About', () => {
         await expect(AboutScreen.databaseSchemaVersionTitle).toHaveText('Database Schema Version:');
         await expect(AboutScreen.databaseSchemaVersionValue).toBeVisible();
         await expect(AboutScreen.copyInfoButton).toBeVisible();
-        await expect(AboutScreen.copyInfoButton).toHaveText('Copy info');
```
Comment (larkox) [inline detox/e2e/test/account/about.e2e.ts:72]: Why we don't check the text any more?

File: detox/e2e/test/autocomplete/edit_post.e2e.ts [hunk 6]
```diff
@@ -67,7 +67,7 @@ describe('Autocomplete - Edit Post', () => {
         await expect(Autocomplete.sectionAtMentionList).not.toBeVisible();
 
         // # Type in "@" to activate at-mention autocomplete
-        await EditPostScreen.messageInput.typeText('@');
+        await EditPostScreen.messageInput.typeText(' @');
```
Comment (larkox) [inline detox/e2e/test/autocomplete/edit_post.e2e.ts:70]: Why do we need a space?

File: detox/e2e/test/teams/invite_people.e2e.ts [hunk 7]
```diff
@@ -218,31 +215,31 @@ describe('Teams - Invite', () => {
...
         // * Validate summary
-        await waitFor(Invite.screenSummary).toBeVisible().withTimeout(timeouts.TEN_SEC);
+        waitFor(Invite.screenSummary).toBeVisible();
 
         // * Validate summary report not sent
         await expect(Invite.getSummaryReportNotSent()).toBeVisible();
-        await expect(Invite.getSummaryReportUserItem(testUser1.id)).toBeVisible();
-        await expect(Invite.getSummaryReportUserItemText(testUser1.id)).toHaveText(username1);
+        await expect(Invite.getSummaryReportUserItem(testUser.id)).toBeVisible();
+        await expect(Invite.getSummaryReportUserItemText(testUser.id)).toBeVisible();
```
Comment (larkox) [inline detox/e2e/test/teams/invite_people.e2e.ts:238]: Why don't we check for the text anymore?

File: detox/package.json [hunk 8]
```diff
@@ -20,7 +20,7 @@
     "babel-plugin-module-resolver": "5.0.0",
     "client-oauth2": "4.3.3",
     "deepmerge": "4.3.1",
-    "detox": "20.11.4",
+    "detox": "^20.19.3",
```
Comment (larkox) [inline detox/package.json:23]: Let's try to keep fixed versions.
```suggestion
    "detox": "20.19.3",
```

File: detox/utils/report.js [hunk 9]
```diff
@@ -184,7 +184,7 @@ function removeOldGeneratedReports() {
     ].forEach((file) => fse.removeSync(`${ARTIFACTS_DIR}/${file}`));
 }
 
-function writeJsonToFile(jsonObject, filename, dir) {
+async function writeJsonToFile(jsonObject, filename, dir) {
     fse.writeJson(`${dir}/${filename}`, jsonObject).
         then(() => console.log('Successfully written:', filename)).
         catch((err) => console.error(err));
```
Comment (larkox) [inline detox/utils/report.js:190]: Not sure what you want to achieve with this, but shouldn't be working. Even if you make this function async, it will not await for the result of `fse.writeJson`, since that is in a `promise.then().catch()` pattern, which wouldn't stop execution.

You would have to rewrite this as:
```suggestion
async function writeJsonToFile(jsonObject, filename, dir) {
    try {
        await fse.writeJson(`${dir}/${filename}`, jsonObject);
        console.log('Successfully written:', filename);
    } catch (err) {
        console.error(err);
    }
```

**Review by larkox — COMMENTED** [review]
Not very proficient in the github actions part, so not sure how much help I give there, but here are a few comments about the tests.

File: detox/e2e/support/test_config.ts (hunk 2 above)
Comment (yasserfaraazkhan) [reply detox/e2e/support/test_config.ts:13]: Ohh yes, 😅 
I'll replace these and add to secrets

File: detox/utils/report.js (hunk 9 above)
Comment (yasserfaraazkhan) [reply detox/utils/report.js:190]: Noted Daniel.

File: detox/e2e/test/account/about.e2e.ts (hunk 5 above)
Comment (yasserfaraazkhan) [reply detox/e2e/test/account/about.e2e.ts:72]: this button has icon, the matcher was failing.
I've added other way of asserting text now.

File: detox/e2e/support/ui/screen/login.ts (hunk 4 above)
Comment (yasserfaraazkhan) [reply detox/e2e/support/ui/screen/login.ts:66]: This was a bit weird.
earlier `this.passwordInput.replaceText` was used.
typeText is how user will type using keyboard and it will also trigger input callbacks.

the `\n` was doing the trick of pressing on `next` button, else the keyboard covers elements, and it was extra step of handlign keyboard that I avoided.

File: detox/e2e/support/ui/screen/create_or_edit_channel.ts (hunk 3 above)
Comment (yasserfaraazkhan) [reply detox/e2e/support/ui/screen/create_or_edit_channel.ts:63]: in the smaller 15 Pro, the View Info >  Edit channel option is in bottom and we had to scroll and access.

File: detox/e2e/test/autocomplete/edit_post.e2e.ts (hunk 6 above)
Comment (yasserfaraazkhan) [reply detox/e2e/test/autocomplete/edit_post.e2e.ts:70]: removed

**Review by saturninoabril — CHANGES_REQUESTED** [review]
Thanks @yasserfaraazkhan, great to see all the work you're doing! Overall looks good to me except for few requests.

File: .github/actions/generate-specs/split-tests.js [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    this.directory = directory;
+    this.parallelism = parallelism;
+    this.rawFiles = [];
+    this.groupedFiles = [];
+    this.deviceInfo = deviceInfo;
+  }
+
+  findFiles() {
+    const fs = require('fs');
+    const path = require('path');
```
Comment (saturninoabril) [inline .github/actions/generate-specs/split-tests.js:29]: Let's move this on top.

File: .github/actions/generate-specs/split-tests.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    const dirPath = path.join(this.directory, this.searchPath);
```
Comment (saturninoabril) [inline .github/actions/generate-specs/split-tests.js:30]: Since this is being joined and I don't see them used separately, instead of having two separate env's (`DIRECTORY` and `SEARCH_PATH`), can we use one env only? Maybe `TEST_DIRECTORY` with default value of `detox/e2e/test`?

**Review by saturninoabril — CHANGES_REQUESTED** [review]
Thanks for updating @yasserfaraazkhan! Added new comments.

File: .github/workflows/e2e-detox.yml [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+
+          echo "BUILD_ID=$(echo ${BUILD_ID} | sed 's/ /_/g')" >> $GITHUB_ENV
+          echo "DEVICE_NAME=${DEVICE_NAME}" >> $GITHUB_ENV
+          echo "DEVICE_OS_VERSION=${DEVICE_OS_VERSION}" >> $GITHUB_ENV
+
+      - name: Generate Test Specs
+        id: generate-specs
+        uses: ./.github/actions/generate-specs
+        with:
+          parallelism: 10
```
[the remark this answers is not in the document]
Comment (yasserfaraazkhan) [reply .github/workflows/e2e-detox.yml:62]: I'll try out and put it in readme

**Review by larkox — APPROVED** [review]

File: .github/workflows/e2e-detox.yml [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+env:
+  AWS_REGION: 'us-east-1'
+  ADMIN_USERNAME: ${{ secrets.MM_MOBILE_E2E_ADMIN_EMAIL }}
+  ADMIN_PASSWORD: ${{ secrets.MM_MOBILE_E2E_ADMIN_PASSWORD }}
+  BRANCH: ${{ github.head_ref || github.ref_name }} 
+  COMMIT_HASH: ${{ github.sha }}
+  DETOX_AWS_S3_BUCKET: 'mattermost-detox-report'
+  DETOX_AWS_ACCESS_KEY_ID: ${{ secrets.MM_MOBILE_DETOX_AWS_ACCESS_KEY_ID }}
+  DETOX_AWS_SECRET_ACCESS_KEY: ${{ secrets.MM_MOBILE_DETOX_AWS_SECRET_ACCESS_KEY }}
+  HEADLESS: 'false'
```
Comment (saturninoabril) [inline .github/workflows/e2e-detox.yml:27]: Should this be true?

File: .github/workflows/e2e-detox.yml [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+
+      - name: Prepare iOS Build
+        uses: ./.github/actions/prepare-ios-build
+
+      - name: Build iOS Simulator
+        env:
+          TAG: "${{ github.ref_name }}"
+          AWS_ACCESS_KEY_ID: "${{ secrets.MM_MOBILE_BETA_AWS_ACCESS_KEY_ID }}"
+          AWS_SECRET_ACCESS_KEY: "${{ secrets.MM_MOBILE_BETA_AWS_SECRET_ACCESS_KEY }}"
+          GITHUB_TOKEN: "${{ secrets.MM_MOBILE_GITHUB_TOKEN }}"
```
Comment (saturninoabril) [inline .github/workflows/e2e-detox.yml:89]: Either use `${{ github.token }}` or with the org secret `${{ secrets.GITHUB_TOKEN }}`

File: .github/workflows/e2e-detox.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (saturninoabril) [reply .github/workflows/e2e-detox.yml:69]: Why 10 and not more/less than 10? Does parallelism of `10` give as a good spot between billable cost and acceptable overall duration?

File: detox/e2e/support/test_config.ts [hunk 15]
```diff
@@ -1,14 +1,14 @@
...
-export const siteThreeUrl = process.env.SITE_3_URL || 'https://mobile03.test.mattermost.cloud';
+export const serverOneUrl = process.env.SITE_1_URL || (process.env.IOS === 'true' ? 'http://localhost:8065' : 'http://10.0.2.2:8065');
+export const siteOneUrl = process.env.SITE_1_URL || 'http://localhost:8065';
+export const serverTwoUrl = process.env.SITE_2_URL || 'http://localhost:8065';
+export const siteTwoUrl = process.env.SITE_2_URL || 'http://localhost:8065';
+export const serverThreeUrl = process.env.SITE_3_URL || 'http://localhost:8065';
+export const siteThreeUrl = process.env.SITE_3_URL || 'http://localhost:8065';
 export const smtpUrl = process.env.SMTP_URL || 'http://127.0.0.1:9001';
-export const adminEmail = process.env.ADMIN_EMAIL || 'sysadmin@sample.mattermost.com';
+export const adminEmail = process.env.ADMIN_EMAIL || 'success+sysadmin@simulator.amazonses.com';
```
Comment (saturninoabril) [inline detox/e2e/support/test_config.ts:11]: Since we are using local dev for default setup. I think this should match the ones auto-generated from local. Should this be reverted to `sysadmin@sample.mattermost.com`?

File: detox/package.json [hunk 16]
```diff
@@ -14,13 +14,14 @@
     "@types/jest": "29.5.5",
     "@types/tough-cookie": "4.0.3",
     "@types/uuid": "9.0.4",
+    "async": "^3.2.5",
```
Comment (saturninoabril) [inline detox/package.json:17]: Was this use somewhere? If yes, let's use fix version.
```suggestion
    "async": "3.2.5",
```

**Review by saturninoabril — APPROVED** [review]
Thanks @yasserfaraazkhan, great to see this! All looks good to me except for some comments/suggestions. Approving given those will be resolved.

File: .github/actions/prepare-node-deps/action.yaml [hunk 17]
```diff
@@ -21,6 +21,14 @@ runs:
         node node_modules/\@sentry/cli/scripts/install.js
         echo "::endgroup::"
 
+    - name: Cache Node.js modules
+      uses: actions/cache@0c45773b623bea8c8e75f6c82b208c3cf94ea4f9 # v4.0.2
+      with:
+        path: ~/.npm
+        key: ${{ runner.OS }}-node-${{ hashFiles('**/package-lock.json') }}
+        restore-keys: |
+          ${{ runner.OS }}-node-
```
Comment (mvitale1989) [inline .github/actions/prepare-node-deps/action.yaml:30]:
```suggestion
        key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
        restore-keys: |
          ${{ runner.os }}-node-
```
afaik Github is not case independent, when it comes to context access

File: detox/e2e/support/test_config.ts [hunk 18]
```diff
@@ -1,14 +1,14 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 
-export const serverOneUrl = process.env.SITE_1_URL || (process.env.IOS === 'true' ? 'http://127.0.0.1:8065' : 'http://10.0.2.2:8065');
-export const siteOneUrl = process.env.SITE_1_URL || 'http://127.0.0.1:8065';
-export const serverTwoUrl = process.env.SITE_2_URL || 'https://mobile02.test.mattermost.cloud';
-export const siteTwoUrl = process.env.SITE_2_URL || 'https://mobile02.test.mattermost.cloud';
-export const serverThreeUrl = process.env.SITE_3_URL || 'https://mobile03.test.mattermost.cloud';
-export const siteThreeUrl = process.env.SITE_3_URL || 'https://mobile03.test.mattermost.cloud';
+export const serverOneUrl = process.env.SITE_1_URL || (process.env.IOS === 'true' ? 'http://localhost:8065' : 'http://10.0.2.2:8065');
```
Comment (mvitale1989) [inline detox/e2e/support/test_config.ts:4]: Is the `http://10.0.2.2:8065` fallback still required?

**Review by mvitale1989 — COMMENTED** [review]
I have left some inital comments and questions, I'll take a deeper look at the workflow files on Monday

File: detox/e2e/support/test_config.ts (hunk 18 above)
Comment (yasserfaraazkhan) [reply detox/e2e/support/test_config.ts:4]: yes, to run locally (server + emulator) we do need to point it to this.

File: .github/workflows/e2e-detox.yml [hunk 19]
```diff
@@ -0,0 +1,214 @@
+name: Detox E2E Tests
+
+on:
+  push:
+    branches:
+      - main
```
[the remark this answers is not in the document]
Comment (mvitale1989) [reply .github/workflows/e2e-detox.yml:6]: Coming back on this, one additional consideration is that approach risks hiking our CI costs, since each merge to main will trigger a full E2E run.
On the monorepo, instead of running the E2E test on each merge to the `main` branch, we instead have a daily E2E run against the master branch.

I would propose removing the `main` branch from the `on.push` trigger, and create a `schedule` trigger instead, for testing the main branch regularly.

**Comment by yasserfaraazkhan** [comment]
Hi @mvitale1989 
should these changes be okay to merge ?

File: .github/workflows/e2e-detox-ci-template.yml [hunk 20]
```diff
@@ -0,0 +1,299 @@
...
+          name: ios-build-simulator-${{ github.run_id }}
+          path: Mattermost-simulator-x86_64.app.zip
+
+  e2e-ios:
+    continue-on-error: true
+    env:
+      IOS: true
+    runs-on: macos-14-large
+    name: ios-detox-e2e-${{ matrix.runId }}-${{ matrix.deviceName }}-${{ matrix.deviceOsVersion }}
+    if: ${{ inputs.run-ios-tests }}
```
Comment (mvitale1989) [inline .github/workflows/e2e-detox-ci-template.yml:139]:
```suggestion
    if: ${{ inputs.run-ios-tests }}
    timeout-minutes: 40
```
This is something that bit us in the monorepo :smile: since these are expensive instances, it's better to limit their runtime to a maximum.

**Review by mvitale1989 — CHANGES_REQUESTED** [review]
Thanks Yasser! I only see one last important point that would needs implementing before merging, everything else that may be improved (like the many status checks) we can do it afterwards :+1:

**Review by mvitale1989 — APPROVED** [review]
